platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/service/WritableRosterStore.java

```diff
@@ -0,0 +1,231 @@
+     */
+    @Nullable
+    @Override
+    public Roster getCandidateRoster() {
+        return rosterStateAccessor.getCandidateRoster();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @NonNull
+    @Override
+    public Roster getActiveRoster() {
+        return rosterStateAccessor.getActiveRoster();
+    }
+
+    /**
+     * Adopts the candidate roster present in the state as the active roster.
+     * Until the Dynamic Address Book is implemented, this method will be called unconditionally on network upgrade.
+     *
+     * @param roundNumber the round number in which the candidate roster became active
+     */
+    private void adoptCandidateRoster(final long roundNumber) {
+        final RosterState previousRosterState = rosterStateOrThrow();
+        final Roster candidateRoster = rosterMap.get(ProtoBytes.newBuilder()
+                .value(previousRosterState.candidateRosterHash())
+                .build());
+        if (candidateRoster == null) {
+            throw new IllegalStateException("Candidate roster not found in the state.");
+        }
+
+        storeAsActive(candidateRoster, roundNumber);
+    }
+
+    /**
+     * Stores this roster as the active roster.
+     *
+     * @param roster        a roster to set as active
+     * @param round        the round in which this roster became active
+     */
+    private void storeAsActive(@NonNull final Roster roster, final long round) {
+        Objects.requireNonNull(roster);
+        RosterValidator.validate(roster);
+
+        final RosterState previousRosterState = rosterStateOrThrow();
+
+        // update the roster state
+        final List<RoundRosterPair> roundRosterPairs = new LinkedList<>(previousRosterState.roundRosterPairs());
+        final Bytes activeRosterHash = RosterUtils.hash(roster).getBytes();
+        roundRosterPairs.addFirst(new RoundRosterPair(round, activeRosterHash));
+
+        // remove the formerly previous active roster, i.e., the roster that was active before the last two adopted
+        // rosters, if any.
+        if (roundRosterPairs.size() > 2) {
+            final RoundRosterPair lastRemovedStillInState = roundRosterPairs.removeLast();
+
+            // remove the roster from the map if it is still present in the Roster state for whatever reason.
+            final boolean rosterStateDuplicateFound = roundRosterPairs.stream()
+                    .anyMatch(peer -> peer.activeRosterHash().equals(lastRemovedStillInState.activeRosterHash()));
+
+            if (!rosterStateDuplicateFound) {
+                this.rosterMap.remove(ProtoBytes.newBuilder()
+                        .value(lastRemovedStillInState.activeRosterHash())
+                        .build());
+            }
+
+            // At this phase of the implementation,
+            // the roster state should not contain more than 3 active rosters at most.
+            // Future implementations (e.g. DAB) can modify this.
+            if (roundRosterPairs.size() > 3) {
```

Comment:
It's already checking for size >2 and removing last to get it back down to 2. I think it won't be greater than 2 ever considering the removal on line 186, which is expected to be hit often, and will get the size back to 2.
This check on line 201 for >3 is additional insurance, which shouldn't really ever be hit. 

The alternative is to remove before add, but i think it's safer to add-then-remove.

Reply:
With the current code, it is impossible for the `roundRosterPairs` to be length 3 or greater at this line. So if you want an additional safety check, it should be `> 2` of `>= 3` instead of `> 3`

Reply:
> The alternative is to remove before add, but i think it's safer to add-then-remove.

I'm not saying the order of list operations need to change. Just that we should assert the list size is within the expected bounds, at the correct place.

> but given the logic you've implemented, I don't think that is necessary.

Since there is no testing which checks that there isn't a logic flaw that causes the list to grow indefinitely, I think a max size safety check is warranted.

Reply:
@litt3 I meant the check using a value of 3 instead of 2